File: second course/PDOO/civitasRuby/lib/titulo_propiedad.rb
# encoding:utf-8

module Civitas
	class Titulo_propiedad
		
		attr_reader   :num_casas, :num_hoteles, :precio_compra,
                  :precio_edificar, :propietario, :hipotecado, :nombre
		
		@FACTOR_INTERESES_HIPOTECA = 1.1
		
		def initialize(nom, ab, fr, hb, pc, pe)
			@nombre = nom
			@alquiler_base = ab
			@factor_revalorizacion = fr
			@hipoteca_base = hb
			@precio_compra = pc
			@precio_edificar = pe
			@propietario = nil
			@num_casas = 0
			@num_hoteles = 0
			@hipotecado = false 
		end
		
		private
		
		def precio_alquiler()
			precio_alquiler = 0 #inicializando al caso de estar hipotecado
			                    #o esta el propietario encarcelado
													
			if(!propietario_encarcelado && !@hipotecado)
					precio_alquiler = @alquiler_base * (1 + (@num_casas * 0.5) + (@num_hoteles * 2.5))
			end
			
			return precio_alquiler
		end
		
		def propietario_encarcelado()
			encarcelado = true
			
			if(!@propietario.encarcelado() || !tienePropietario())
				encarcelado = false
			end
			
			return encarcelado
		end
		
		def precio_venta()
			return @precio_compra + (cantidad_casas_hoteles() * @precio_edificar) * @factor_revalorizacion
		end
		
		def es_este_el_propietario(jugador)
			return jugador == propietario
		end
		
		def importe_hipoteca()
			return @hipoteca_base * (1 + (@num_casas * 0.5) + (@num_hoteles * 2.5))
		end
		
		
		public
		
		def to_string()
			return "\n\t\tnombre: #{@nombre}
							\n\t\tAlquiler Base: #{@alquiler_base}
              \n\t\tfactor revalorizacion: #{@factor_revalorizacion} 
              \n\t\thipoteca base: #{@hipoteca_base} 
							\n\t\thipotecado: #{@hip}
              \n\t\tnºcasas: #{@num_casas}  
							\n\t\tnºhoteles: #{@num_hoteles} 
              \n\t\tprecio compra: #{@precio_compra}
              \n\t\tprecio edificar: #{@precio_edificar}"
		end
		
		def importe_cancelar_hipoteca()
			cantidad_recibida = @hipoteca_base * (1 + (@num_casas * 0.5) + (@num_hoteles * 2.5))
			
			return cantidad_recibida * @FACTOR_INTERESES_HIPOTECA
		end
		
		def tramitar_alquiler(jugador)
			if(tiene_propietario() && !es_este_el_propietario(jugador))
				alquiler = precio_alquiler()
				
				jugador.paga_alquiler(alquiler)
				@propietario.recibe(alquiler)
			end
			
		end
		
		def cantidad_casas_hoteles()
			return @num_casas + @num_hoteles
		end
		
		def derruir_casas(n, jugador)
			casa_derruida = false
			
			if(es_este_el_propietario(jugador) && @num_casas >= n)
				@num_casas -= n
				casa_derruida = true
			end
			
			return casa_derruida
		end
		
		def vender(jugador)
			vendido = false
			
			if(es_este_el_propietario(jugador) && !@hipotecado)
				jugador.recibe(precio_venta())
				@num_casas = 0
				@num_hoteles = 0
				@propietario = nil
				vendido = true
			end
			
			return vendido
		end
		
		def actualiza_propietario_por_conversion(jugador)
			@propietario = jugador
		end
		
		def hipotecar(jugador)
			salida = false
			
			if(!@hipotecado && es_este_el_propietario(jugador))
				jugador.recibe(importe_hipoteca())
				@hipotecado = true
				salida = true
			end
			
			return salida
		end
		
		def cancelar_hipoteca(jugador)
			result = false
			
			if(@hipotecado && es_este_el_propietario(jugador))
				@propietario.paga(importe_cancelar_hipoteca())
				@hipotecado = false
				result = true
			end
			
			return result
		end
		
		def comprar(jugador)
			result = false
			
			if(!tiene_propietario())
				@propietario = jugador
				result = true
				@propietario.paga(@precio_compra)
			end
			
			return result
		end
		
		def construir_casa(jugador)
			result = false
			
			if(es_este_el_propietario(jugador))
				result = true
				@num_casas += 1
				jugador.paga(@precio_edificar)
			end
			
			return result
		end
		
		def construir_hotel(jugador)
			result = false
			
			if(es_este_el_propietario(jugador))
				jugador.paga(@precio_edificar)
				@num_hoteles += 1
				result = true
			end
			
			return result
		end
		
		def importe_hipoteca()
			return @hipoteca_base * (1 + (@num_casas * 0.5) + (@num_hoteles * 2.5))
		end
		
		def tiene_propietario()
			return propietario != nil
		end
		
		

	end	
end

File: second course/PDOO/civitasJava/src/civitas/TituloPropiedad.java
package civitas;

public class TituloPropiedad{
  
  private static float factorInteresesHipoteca = 1.1f;
  
  private float alquilerBase;
  private float factorRevalorizacion;
  private float hipotecaBase;
  private boolean hipotecado;
  private String nombre;
  private int numCasas;
  private int numHoteles;
  private float precioCompra;
  private float precioEdificar;
  
  //Atributos de referencia
  private Jugador propietario;
  
  
  TituloPropiedad(String nom, float ab, float fr, float hb, float pc, float pe){ //TESTED
		nombre = nom;
		alquilerBase = ab;
		factorRevalorizacion = fr;
		hipotecaBase = hb;
		precioCompra = pc;
		precioEdificar = pe;
		propietario = null;
		numCasas = 0;
		numHoteles = 0;
		hipotecado = false;
  }
  
  @Override
  public String toString(){ //TESTED
    return "\n\t\tnombre: " + nombre + "\n\t\tAlquiler Base: " + alquilerBase
            + "\n\t\tfactor revalorizacion: " + factorRevalorizacion +
            "\n\t\thipoteca base: " + hipotecaBase + "\n\t\thipotecado: " + hipotecado
            + "\n\t\tnºcasas: " + numCasas + "\n\t\tnºhoteles: " + numHoteles +
            "\n\t\tprecio compra: "+ precioCompra +"\n\t\tprecio edificar: "
            + precioEdificar;
  }
  
  private float getPrecioAlquiler(){ //TESTED
		float precioAlquiler = 0; //inicializado al caso de estar hipotecado
                              //o esta el propietario encarcelado
							  
		if(!propietarioEncarcelado() && !hipotecado)
			precioAlquiler = alquilerBase * (1 + (numCasas * 0.5f) + (numHoteles * 2.5f));
		
		return precioAlquiler;
  }
  
  float getImporteCancelarHipoteca(){
		float cantidadRecibida = hipotecaBase * (1 + (numCasas * 0.5f) + (numHoteles * 2.5f));
	
		return cantidadRecibida * factorInteresesHipoteca;
  }
  
  void tramitarAlquiler(Jugador jugador){ //TESTED
		if(tienePropietario() && !esEsteElPropietario(jugador)){
			float alquiler = getPrecioAlquiler();
	  
			jugador.pagaAlquiler(alquiler);
			propietario.recibe(alquiler);
		}
  }
  
  private boolean propietarioEncarcelado(){ //TESTED
		boolean encarcelado = true;
	
		if(!propietario.isEncarcelado() || !tienePropietario())
			encarcelado = false;
	
		return encarcelado;
  }
  
  int cantidadCasasHoteles(){ //TESTED
		return numCasas + numHoteles;
  }
  
  private float getPrecioVenta(){ //TESTED
    return precioCompra + (cantidadCasasHoteles() * precioEdificar) * factorRevalorizacion;
  }
  
  boolean derruirCasas(int n, Jugador jugador){ //TESTED
		boolean casasDerruidas = false;
	
		if(esEsteElPropietario(jugador) && numCasas >= n){
			numCasas -= n;
			casasDerruidas = true;
		}  
	  
		return casasDerruidas;  
  }
  
  boolean vender(Jugador jugador){ //TESTED
	
		boolean vendido = false;
	
		if(esEsteElPropietario(jugador) && !hipotecado){
			jugador.recibe(getPrecioVenta());
			numCasas = 0;
			numHoteles = 0;
			propietario = null;
			vendido = true;
		}  
	
		return vendido;
  }
  
  void actualizaPropietarioPorConversion(Jugador jugador){ //TESTED
		propietario = jugador;
  }
  
  
  boolean hipotecar(Jugador jugador){ //TESTED
		boolean salida = false;
	
		if(!hipotecado && esEsteElPropietario(jugador)){
			jugador.recibe(getImporteHipoteca());
			hipotecado = true;
			salida = true;
		}
	
		return salida;
  }
 
  boolean cancelarHipoteca(Jugador jugador){ //TESTED
		boolean result = false;
	
		if(hipotecado && esEsteElPropietario(jugador)){
			propietario.paga(getImporteCancelarHipoteca());
			hipotecado = false;
			result = true;
		}
	
		return result;
  }
  
  boolean comprar(Jugador jugador){ //TESTED
		boolean result = false;
	
		if(!tienePropietario()){
			propietario = jugador;
			result = true;
			propietario.paga(precioCompra);
		}
	
		return result;
  }
  
  boolean construirCasa(Jugador jugador){ //TESTED???
		boolean result = false;
    
		if(esEsteElPropietario(jugador)){
      result = true;
      numCasas++;
    }
        
		return result;
  }
  
  boolean construirHotel(Jugador jugador){ //TESTED
		boolean result = false;
	
		if(esEsteElPropietario(jugador)){
			jugador.paga(precioEdificar);
			numHoteles++;
			result = true;
		}
	
		return result;
  }
  
  private boolean esEsteElPropietario(Jugador jugador){ //TESTED
		return jugador == propietario;
  }
  
  public boolean getHipotecado(){
		return hipotecado;
  }
  
  private float getImporteHipoteca(){ //TESTED
		return hipotecaBase * (1 + (numCasas * 0.5f) + (numHoteles * 2.5f));
  }
  
	//VISIBILIDAD CAMBIADA PARA PODER USARLO EN EL OTRO MODULO
  public String getNombre(){
		return nombre;
  }
  
  int getNumCasas(){
		return numCasas;
  }
  
  int getNumHoteles(){
		return numHoteles;
  }
  
  float getPrecioCompra(){
		return precioCompra;
  }
  
  float getPrecioEdificar(){
		return precioEdificar;
  }
  
  Jugador getPropietario(){
		return propietario;
  }
  
  boolean tienePropietario(){ //TESTED
		return propietario != null;
  }
	
	public static void main(String [] args){
		TituloPropiedad t = new TituloPropiedad("Calle Madrid", 30.36f, 45.6f, 23.4f, 14.6f, 43.6f);
		
		Jugador j1 = new Jugador("Paco");
		Jugador j2 = new Jugador("Juan");
		
		t.comprar(j1);
		
		t.construirCasa(j1);
		t.construirCasa(j1);
		t.construirCasa(j1);
		t.construirCasa(j1);
		
		t.derruirCasas(3, j1);
		
		while(Diario.getInstance().eventosPendientes())
			System.out.println(Diario.getInstance().leerEvento());
	}
}
